### DISCUSSION TITLE
React JS Tutorial: Building a Recipe Search App using Functional Components

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/mern#181] [Tasks]
In recent years, the trend in React development has moved towards functional components, largely due to the introduction of hooks, which allow you to use state and other React features without writing a class.

Let's walk through the creation of the same "Recipe Search App" using functional components and hooks. 

**Step 1: Project Setup**

We'll begin by creating a new React project using Create React App.

```bash
npx create-react-app recipe-search
cd recipe-search
npm start
```

**Step 2: App Component**

`App.js` serves as the main component for our application. We're going to import our other components into this file.

```jsx
import React from 'react';
import SearchBar from './SearchBar';
import RecipeList from './RecipeList';

function App() {
  return (
    <div>
      <h1>Recipe Search App</h1>
      <SearchBar />
      <RecipeList />
    </div>
  );
}

export default App;
```

**Step 3: SearchBar Component**

Let's create the `SearchBar` component with a text field for users to enter their search term and a button to trigger the search.

```jsx
import React from 'react';

function SearchBar() {
  return (
    <div>
      <input type="text" placeholder="Search for recipes" />
      <button>Search</button>
    </div>
  );
}

export default SearchBar;
```

**Step 4: RecipeList Component**

Next, we create a `RecipeList` component that will later display the search results.

```jsx
import React from 'react';

function RecipeList() {
  return <div>Recipe List goes here</div>;
}

export default RecipeList;
```

**Step 5: Using State with Hooks**

State is a built-in React object used to contain data or information about the component. In functional components, we use the `useState` hook to create a state variable. Let's modify `App` to include state for the search keyword and the recipe list:

```jsx
import React, { useState } from 'react';
// ...

function App() {
  const [keyword, setKeyword] = useState('');
  const [recipes, setRecipes] = useState([]);

  // ...
}
```

In the example above, `useState('')` and `useState([])` initialize `keyword` and `recipes` with an empty string and an empty array, respectively. The `useState` hook returns a pair: the current state value and a function that lets you update it. 

Think of `useState` as a way to "borrow" statefulness from React. It's like renting a storage locker: you ask for some space (declaring a state variable), and in return, you get a key (the function) that lets you put things in and take things out.

**Step 6: Passing Props**

Props, short for properties, allow us to pass values from a parent component down to a child component. Let's pass our state down to the `RecipeList` component:

```jsx
<RecipeList recipes={recipes} />
```

In `RecipeList`, we can then access `recipes` like this:

```jsx
function RecipeList({ recipes }) {
  return <div>{recipes.length} recipes found</div>;
}
```

**Step 7: Handling Events**

Next, let's update our `SearchBar` component to handle user input and button click events. We use the `useState` hook to manage the text input's state:

```jsx
import React, { useState } from 'react';

function SearchBar({ onSearch }) {
  const [input, setInput] = useState('');

  const handleInputChange = (event) => {
    setInput(event.target.value);
  };

  const handleSearchClick = () => {
    onSearch(input);
  };

  return (
    <div>
      <input type="text" onChange={handleInputChange} placeholder="Search for recipes" />
      <button onClick={handleSearchClick}>Search</button>
    </div>
  );
}

export default SearchBar;
```

In the `SearchBar` component, we added two new functions: `handleInputChange` and `handleSearchClick`. These functions handle the change of the input field and the click of the search button, respectively.

**Step 8: Fetching Data with useEffect**

The `useEffect` hook lets you perform side effects in function components. In our case, we'll use it to fetch data from the API whenever the `keyword` state changes. This hook will be added to the `App` component:

```jsx
import React, { useState, useEffect } from 'react';
// ...

function App() {
  const [keyword, setKeyword] = useState('');
  const [recipes, setRecipes] = useState([]);

  useEffect(() => {
    const fetchData = async () => {
      const response = await fetch(`https://api.example.com/recipes?keyword=${keyword}`);
      const data = await response.json();
      setRecipes(data.results);
    };

    if (keyword) {
      fetchData();
    }
  }, [keyword]);

  const handleSearch = (searchTerm) => {
    setKeyword(searchTerm);
  };

  return (
    <div>
      <h1>Recipe Search App</h1>
      <SearchBar onSearch={handleSearch} />
      <RecipeList recipes={recipes} />
    </div>
  );
}

export default App;
```

The `useEffect` hook accepts two arguments: a function that contains the effect (fetching data in our case) and a dependency array. The effect will only run when the values in the dependency array (`keyword` in our case) change.

**Step 9: Rendering the Recipes**

Finally, we render our recipes in the `RecipeList` component:

```jsx
function RecipeList({ recipes }) {
  return (
    <div>
      {recipes.map((recipe) => (
        <div key={recipe.id}>
          <h2>{recipe.title}</h2>
          <img src={recipe.image} alt={recipe.title} />
          <p>{recipe.description}</p>
        </div>
      ))}
    </div>
  );
}
```

**Comment by starvikas1988** [comment]
<img width="946" alt="image" src="https://github.com/akash-coded/mern/assets/70792693/65a7a6f0-2daf-40c5-ba62-251ff3097e4b">

**Comment by abhijitjd003** [comment]
![image](https://github.com/akash-coded/mern/assets/34028704/521d5d34-c931-45a6-813a-25a19caac4c0)

**Comment by abhijitjd003** [comment]
![image](https://github.com/akash-coded/mern/assets/34028704/dc18a7e3-271b-4863-bf26-eb9a8a00497a)

**Comment by RevathiSundaramoorthy** [comment]
Used reqres.in instead of apiexample.com/recipes
#### app.js
```javascript

  useEffect(() => {
    const fetchData = async () => {
      const response = await fetch(`https://reqres.in/api/unknown`);
      const data = await response.json();
      setRecipes(data.data);
    };
```
#### recipes.js
```javascript
function RecipeList({ recipes }) {
    return (
        <div>
          {recipes.map((recipe) => (
            <div key={recipe.id}>
              <h2>{recipe.name}</h2>
              <p>{recipe.color}</p>
            </div>
          ))}
        </div>
      );
  }
```
![image](https://github.com/akash-coded/mern/assets/126975777/932f6138-81c0-4dd6-8a18-eb00719819ee)

**Comment by sam2elect** [comment]
![image](https://github.com/akash-coded/mern/assets/85131631/ad287222-0c75-4db5-b1e7-88498706ba89)

**Comment by basu-cse** [comment]
![search](https://github.com/akash-coded/mern/assets/95092878/55966cdb-c7a0-49bc-9d97-be74aee27c77)

**Comment by mownica-p** [comment]
![image](https://github.com/akash-coded/mern/assets/126975660/431a5109-9058-4061-80c2-403bf75f51ec)

**Comment by vk0027** [comment]
![image](https://github.com/akash-coded/mern/assets/123943655/ba78a706-7fcd-42ff-b2ee-a6bea78be210)
